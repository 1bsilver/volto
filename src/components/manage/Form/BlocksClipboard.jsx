import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { injectIntl } from 'react-intl';
import { Portal } from 'react-portal';
import {
  messages,
  getBlocksFieldname,
  getBlocksLayoutFieldname,
} from '@plone/volto/helpers';
import { Icon } from '@plone/volto/components';
import { v4 as uuid } from 'uuid';
import { load } from 'redux-localstorage-simple';
import { isEqual } from 'lodash';

import { setBlocksClipboard, resetBlocksClipboard } from '@plone/volto/actions';
import { blocks } from '~/config';

import copySVG from '@plone/volto/icons/copy.svg';
import pasteSVG from '@plone/volto/icons/paste.svg';

class BlockClipboardHandler extends React.Component {
  loadFromStorage = () => {
    const blocksData = load({ states: ['blocksClipboard'] })?.blocksClipboard
      ?.blocksData;
    if (!isEqual(blocksData, this.props.blocksClipboard))
      this.props.setBlocksClipboard(blocksData);
  };

  componentDidMount() {
    window.addEventListener('storage', this.loadFromStorage, true);
  }

  componentWillUnmount() {
    window.removeEventListener('storage', this.loadFromStorage);
  }

  copyBlocksToClipboard = () => {
    const { formData } = this.props;
    const blocksFieldname = getBlocksFieldname(formData);
    const blocks = formData[blocksFieldname];
    const blocksData = this.props.selectedBlocks.map(
      (blockId) => blocks[blockId],
    );
    this.props.setBlocksClipboard(blocksData);
    this.props.onSetSelectedBlocks([]);
  };

  pasteBlocks = () => {
    const blocksData = this.props.blocksClipboard;
    const cloneWithIds = blocksData
      .filter((blockData) => !!blockData['@type'])
      .map((blockData) => {
        const blockConfig = blocks.blocksConfig[blockData['@type']];
        return blockConfig.cloneData
          ? blockConfig.cloneData(blockData) // returns [blockId, blockData]
          : [uuid(), blockData];
      })
      .filter((info) => !!info); // some blocks may refuse to be copied
    const { formData } = this.props;
    const blocksFieldname = getBlocksFieldname(formData);
    const blocksLayoutFieldname = getBlocksLayoutFieldname(formData);
    const selectedIndex =
      formData[blocksLayoutFieldname].items.indexOf(this.props.selectedBlock) +
      1;

    const newBlockData = {
      [blocksFieldname]: {
        ...formData[blocksFieldname],
        ...Object.assign(
          {},
          ...cloneWithIds.map(([id, data]) => ({ [id]: data })),
        ),
      },
      [blocksLayoutFieldname]: {
        ...formData[blocksLayoutFieldname],
        items: [
          ...formData[blocksLayoutFieldname].items.slice(0, selectedIndex),
          ...cloneWithIds.map(([id]) => id),
          ...formData[blocksLayoutFieldname].items.slice(selectedIndex),
        ],
      },
    };

    this.props.resetBlocksClipboard();
    this.props.onChangeBlocks(newBlockData);
  };

  render() {
    return (
      <>
        {this.props.selectedBlock && this.props.selectedBlocks.length > 0 ? (
          <Portal
            node={
              __CLIENT__ && document.querySelector('#toolbar .toolbar-actions')
            }
          >
            <button
              aria-label={this.props.intl.formatMessage(messages.copyBlocks)}
              onClick={this.copyBlocksToClipboard}
              tabIndex={0}
              className="copyBlocks"
              id="toolbar-copy-blocks"
            >
              <Icon name={copySVG} size="30px" className="circled" />
            </button>
          </Portal>
        ) : (
          ''
        )}
        {this.props.selectedBlock && this.props.blocksClipboard.length > 0 ? (
          <Portal
            node={
              __CLIENT__ && document.querySelector('#toolbar .toolbar-actions')
            }
          >
            <button
              aria-label={this.props.intl.formatMessage(messages.pasteBlocks)}
              onClick={this.pasteBlocks}
              tabIndex={0}
              className="pasteBlocks"
              id="toolbar-paste-blocks"
            >
              <span class="blockCount">
                {this.props.blocksClipboard.length}
              </span>
              <Icon name={pasteSVG} size="30px" className="circled" />
            </button>
          </Portal>
        ) : (
          ''
        )}
      </>
    );
  }
}

export default compose(
  injectIntl,
  connect(
    (state) => {
      return {
        blocksClipboard: state?.blocksClipboard?.blocksData || [],
      };
    },
    { setBlocksClipboard, resetBlocksClipboard },
  ),
)(BlockClipboardHandler);
