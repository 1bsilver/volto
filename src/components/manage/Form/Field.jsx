/**
 * Field component.
 * @module components/manage/Form/Field
 */

import React, { PropTypes } from 'react';

/**
 * Field component class.
 * @function Field
 * @returns {string} Markup of the component.
 */
const Field = ({ id, classNames, label, help, required, rawDescription, rawErrors, children }) =>
  <div className={`field ${classNames} ${rawErrors.length > 0 ? 'error' : ''}`}>
    <label htmlFor={id} className="horizontal">
      {label}
      {required ? <span className="required horizontal" title="Required">&nbsp;</span> : null}
      {rawDescription ? <span className="formHelp">{rawDescription}</span> : null}
    </label>
    {rawErrors.map(error =>
      <div key={error} className="fieldErrorBox">{error}</div>,
    )}
    {children}
    {help}
  </div>;

/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
Field.propTypes = {
  id: PropTypes.string.isRequired,
  classNames: PropTypes.string.isRequired,
  label: PropTypes.string,
  help: PropTypes.element,
  required: PropTypes.bool,
  rawDescription: PropTypes.string,
  rawErrors: PropTypes.arrayOf(PropTypes.string),
  children: PropTypes.element.isRequired,
};

/**
 * Default properties.
 * @property {Object} defaultProps Default properties.
 * @static
 */
Field.defaultProps = {
  label: '',
  help: null,
  required: false,
  rawDescription: '',
  rawErrors: [],
};

export default Field;
