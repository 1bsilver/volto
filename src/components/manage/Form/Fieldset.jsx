/**
 * Fieldset component.
 * @module components/manage/Form/Fieldset
 */

import React, { PropTypes } from 'react';

/**
 * Fieldset component class.
 * @function Fieldset
 * @returns {string} Markup of the component.
 */
const Fieldset = ({ id, title, active, fields }) =>
  <fieldset
    id={`fieldset-${id}`}
    className={`autotoc-section ${active ? 'active' : ''}`}
  >
    <legend id={`fieldsetlegend-${id}`}>{title}</legend>
    {fields.map(field =>
      <div>{field.title}</div>,
    )}
  </fieldset>;

/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
Fieldset.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  active: PropTypes.bool.isRequired,
  fields: PropTypes.arrayOf(
    PropTypes.object,
  ).isRequired,
};

export default Fieldset;
