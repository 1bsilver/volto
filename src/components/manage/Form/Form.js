/**
 * Form component.
 * @module components/manage/Form/Form
 */

import React from 'react';
import SchemaForm from 'react-jsonschema-form';
import { omitBy } from 'lodash';

import { WysiwygWidget } from '../../../components';

/**
 * Form component class.
 * @function Form
 * @returns {string} Markup of the component.
 */
const Form = ({ schema, formData, onSubmit, onCancel }) => (
  <SchemaForm schema={{
                ...schema,
                required: ['title'],
              }}
              uiSchema={{
                text: {
                  "ui:widget": WysiwygWidget,
                }
              }}
              formData={omitBy(formData, item => item === null)}
              onSubmit={onSubmit}>
    <div className="formControls">
      <button className="context" type="submit">Save</button>
      &nbsp;
      <button type="button" onClick={onCancel}>Cancel</button>
    </div>
  </SchemaForm>
);

export default Form;
