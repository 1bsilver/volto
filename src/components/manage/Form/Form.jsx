/**
 * Form component.
 * @module components/manage/Form/Form
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { map } from 'lodash';
import { Button, Form as UiForm, Menu, Segment } from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';

import { Field } from '../../../components';

/**
 * Form container class.
 * @class Form
 * @extends Component
 */
export default class Form extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    schema: PropTypes.shape({
      fieldsets: PropTypes.arrayOf(
        PropTypes.shape({
          fields: PropTypes.arrayOf(PropTypes.string),
          id: PropTypes.string,
          title: PropTypes.string,
        }),
      ),
      properties: PropTypes.objectOf(PropTypes.any),
      required: PropTypes.arrayOf(PropTypes.string),
    }).isRequired,
    formData: PropTypes.objectOf(PropTypes.any),
    onSubmit: PropTypes.func.isRequired,
    onCancel: PropTypes.func.isRequired,
  };

  /**
   * Default properties.
   * @property {Object} defaultProps Default properties.
   * @static
   */
  static defaultProps = {
    formData: {},
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs WysiwygEditor
   */
  constructor(props) {
    super(props);
    this.state = {
      currentTab: 0,
      formData: props.formData,
    };
    this.selectTab = this.selectTab.bind(this);
    this.onChangeField = this.onChangeField.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /**
   * Change field handler
   * @method onChangeField
   * @param {string} id Id of the field
   * @param {*} value Value of the field
   * @returns {undefined}
   */
  onChangeField(id, value) {
    this.setState({
      formData: {
        ...this.state.formData,
        [id]: value,
      },
    });
  }

  /**
   * Submit handler
   * @method onSubmit
   * @param {Object} event Event object.
   * @returns {undefined}
   */
  onSubmit(event) {
    this.props.onSubmit(this.state.formData);
    event.preventDefault();
  }

  /**
   * Select tab handler
   * @method selectTab
   * @param {Object} event Event object.
   * @param {number} index Selected tab index.
   * @returns {undefined}
   */
  selectTab(event, { index }) {
    this.setState({
      currentTab: index,
    });
  }

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    const { schema, onCancel } = this.props;
    const currentFieldset = schema.fieldsets[this.state.currentTab];

    const fields = map(currentFieldset.fields, field => ({
      ...schema.properties[field],
      id: field,
      value: this.state.formData[field],
      required: schema.required.indexOf(field) !== -1,
      onChange: this.onChangeField,
    }));

    return (
      <UiForm method="post" onSubmit={this.onSubmit}>
        <Menu attached="top" tabular stackable>
          {map(schema.fieldsets, (item, index) => (
            <Menu.Item
              name={item.id}
              index={index}
              key={item.id}
              active={this.state.currentTab === index}
              onClick={this.selectTab}
            >
              {item.title}
            </Menu.Item>
          ))}
        </Menu>
        <Segment attached>
          {fields.map(field => <Field {...field} key={field.id} />)}
        </Segment>
        <Segment attached="bottom">
          <Button primary type="submit">
            <FormattedMessage id="Save" defaultMessage="Save" />
          </Button>
          <Button onClick={onCancel}>
            <FormattedMessage id="Cancel" defaultMessage="Cancel" />
          </Button>
        </Segment>
      </UiForm>
    );
  }
}
