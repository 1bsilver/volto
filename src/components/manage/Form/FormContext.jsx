import React from 'react';

export const FormStateContext = React.createContext();

export const FormStateProvider = (props) => {
  const { initialValue } = props;
  const [contextData, setContextData] = React.useState(initialValue);

  const logger = (val) => {
    setContextData((data) => ({ ...data, ...val }));
  };

  return (
    <FormStateContext.Provider value={{ contextData, setContextData: logger }}>
      {props.children}
    </FormStateContext.Provider>
  );
};

export const useFormStateContext = () => {
  const context = React.useContext(FormStateContext);

  if (!context) {
    throw new Error(
      `The \`useFormStateContext\` hook must be used inside the <FormStateProvider> component's context.`,
    );
  }

  return context;
};
