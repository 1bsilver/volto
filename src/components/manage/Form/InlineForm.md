```jsx
import Wrapper from '@plone/volto/styleguide'

const IframeSchema = {
  title: 'Embed external content',

  fieldsets: [
    {
      id: 'default',
      title: 'Default',
      fields: [
        'url',
        'privacy_statement',
        'enabled',
      ],
    },
  ],

  properties: {
    url: {
      title: 'Embed URL',
    },
    privacy_statement: {
      title: 'Privacy statement',
      description: 'Short notification text',
      widget: 'text',
    },
    enabled: {
      title: 'Use privacy screen?',
      description: 'Enable/disable the privacy protection',
      type: 'boolean',
    },
  },

  required: ['url'],
};

<Wrapper>
  <InlineForm
    title={IframeSchema.title}
    schema={IframeSchema}
    formData={{url: 'https://google.com', enabled: true}}
  />
</Wrapper>
```
