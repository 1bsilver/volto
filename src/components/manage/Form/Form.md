Form example source:

```jsx noeditor
const { Provider } = require('react-intl-redux');
const configureStore = require('redux-mock-store').default;
import BrowserRouter from 'react-router-dom/BrowserRouter';
import Form from './Form';
import StaticRouter from 'react-router-dom/StaticRouter';
const store = configureStore()({
  userSession: {
    login: {},
  },
  intl: {
    locale: 'en',
    messages: {},
  },
  users: {
    reset: {
      loading: '',
    },
  },
  actions: {},
  content: { get: {}, update: {} },
  schema: {},
  controlpanels: {},
});

<div className={'rsg--pre-42'}>
  <Provider store={store}>
    <StaticRouter>
      <Form
        schema={{
          fieldsets: [
            {
              fields: [],
              id: '',
              title: 'PLone',
            },
          ],
          properties: {},
          definitions: {},
          required: [],
        }}
        formData={{}}
        pathname=""
        onSubmit={() => {}}
        onCancel={() => {}}
        submitLabel="PLone"
        resetAfterSubmit={true}
        title={'plone'}
        error={{
          message: 'Hello',
        }}
        loading={true}
        hideActions={true}
        description="CMS"
        visual={true}
        tiles={[]}
      />
    </StaticRouter>
  </Provider>
</div>;
```
