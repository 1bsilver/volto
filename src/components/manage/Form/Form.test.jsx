import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';

import Form from './Form';
import config from '~/config';

const mockStore = configureStore();

jest.mock('./Field', () => jest.fn(() => <div className="Field" />));

describe('Form', () => {
  // config.api = 'Plone';
  it('renders a form component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <Form
          schema={{
            fieldsets: [
              {
                id: 'default',
                title: 'Default',
                fields: ['title'],
              },
            ],
            properties: {
              title: {},
            },
            required: [],
          }}
          onSubmit={() => {}}
          onCancel={() => {}}
        />
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('[guillotina] Form', () => {
  beforeEach(() => {
    config.api = 'guillotina';
  });
  it('asdasd renders a form component', () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <Form
          schema={{
            type_name: 'Document',
            fieldsets: [
              {
                fields: ['title', 'modification_date', 'creation_date', 'text'],
                id: 'default',
                title: 'Default',
              },
              {
                fields: [
                  'guillotina.behaviors.dublincore.IDublinCore|title',
                  'guillotina.behaviors.dublincore.IDublinCore|description',
                  'guillotina.behaviors.dublincore.IDublinCore|creation_date',
                  'guillotina.behaviors.dublincore.IDublinCore|modification_date',
                  'guillotina.behaviors.dublincore.IDublinCore|effective_date',
                  'guillotina.behaviors.dublincore.IDublinCore|expiration_date',
                  'guillotina.behaviors.dublincore.IDublinCore|creators',
                  'guillotina.behaviors.dublincore.IDublinCore|tags',
                  'guillotina.behaviors.dublincore.IDublinCore|publisher',
                  'guillotina.behaviors.dublincore.IDublinCore|contributors',
                ],
                id: 'guillotina.behaviors.dublincore.IDublinCore',
                title: 'Dublin Core fields',
              },
              {
                fields: [
                  'guillotina_cms.interfaces.base.ICMSBehavior|hidden_navigation',
                  'guillotina_cms.interfaces.base.ICMSBehavior|language',
                  'guillotina_cms.interfaces.base.ICMSBehavior|review_state',
                ],
                id: 'guillotina_cms.interfaces.base.ICMSBehavior',
                title: 'CMS data behavior',
              },
              {
                fields: [
                  'guillotina_cms.interfaces.tiles.ITiles|tiles_layout',
                  'guillotina_cms.interfaces.tiles.ITiles|tiles',
                ],
                id: 'guillotina_cms.interfaces.tiles.ITiles',
                title: 'Tiles behavior',
              },
            ],
            properties: {
              title: {},
            },
            definitions: {
              basic: {
                properties: {
                  title: {},
                },
              },
              'guillotina_cms.interfaces.tiles.ITiles': {
                properties: {
                  tiles: {},
                  tiles_layout: [],
                },
              },
            },
            required: [],
          }}
          onSubmit={() => {}}
          onCancel={() => {}}
        />
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
