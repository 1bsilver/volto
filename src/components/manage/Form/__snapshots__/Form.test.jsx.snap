// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders a form component 1`] = `
<form
  className="ui form"
  method="post"
  onSubmit={[Function]}
>
  <div
    className="ui error message"
  >
    <span>
      There were some errors.
    </span>
  </div>
  <div
    className="ui attached segment"
  >
    <div
      className="Field"
    />
  </div>
  <div
    className="ui bottom attached segment"
  >
    <button
      className="ui primary button"
      onClick={[Function]}
      tabIndex={undefined}
      type="submit"
    >
      <span>
        Save
      </span>
    </button>
    <button
      className="ui button"
      onClick={[Function]}
      tabIndex={undefined}
    >
      <span>
        Cancel
      </span>
    </button>
  </div>
</form>
`;
